{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence Analysis: DNA, RNA, and Proteins\n",
    "***\n",
    "\n",
    "## Understanding Biology.\n",
    "Biology studies the evolution, development, physiology, and all molecular interactions & processes that conform living organisms. While biology is a very broad field, **you don't need to be a Biology wizz in order to make new & exciting discoveries.** Perhaps this is mainly due to the fact that every topic in Biology is intrinsically related to what's commonly referred to as **'The Central Dogma'**. **The Central Dogma** describes the flow of information within cells, the basic units of living organisms. This flow of information has been found to be universal amongst contemporary life on Earth, and every physiological phenomena in cells, tissues, and organs can be traced back to individual components present within the Central Dogma.\n",
    "\n",
    "![Alt Text](./Imgs/CentralDogma_Mod.png)\n",
    "> **Figure 1: The Central Dogma.** The genome, the set of instructions and information that encodes for living organisms, consists of **DNA**. Certain portions of DNA, called genes, harbor sequences that are copied into **RNA** throughout a process called **transcription**. Although there are different types of RNA, messenger RNA (mRNA) encodes for polypeptides throughout a process called **translation**. Polypeptides fold and assemble to produce active **protein** products, which in turn carry out the majority of functions and processes that cells need in order to survive.\n",
    "\n",
    "- **DNA** : Harbors genes and regulatory regions that control for the expression of genes. Double-stranded helix consisting of four different types of nucleotides:\n",
    "    - A : Adenine\n",
    "    - C : Cytosine\n",
    "    - G : Guanine\n",
    "    - T : Thyamine\n",
    "\n",
    "- **RNA** : Like DNA, consists of four different types of nucleotides, but is single-stranded. RNA incorporates Uracyl (U), instead of Thyamine (T). Each RNA molecule is assigned a class or 'type' in accordance to its biological role. RNAs that encode for proteins are called messenger RNAs (mRNAs). \n",
    "\n",
    "- **Protein** : Carry out the majority of functions and processes that cells need in order to survive and replicate. Proteins are encoded by **codons**, a sequence of three nucleotides that corresponds with a specific amino acid or stop signal, in RNA (see *Figure 2* below). Can consist of a repertoir of more than 20 amino acids.\n",
    "\n",
    "![Alt Text](./Imgs/TranslationTable.png)\n",
    "> **Figure 2: Codon & Amino acid table.** Graphical representation of codon-amino acid equivalency. To match each codon with an amino acid, follow each nucleotide from the inner regions of the circle to the outer regions of the circle (5'-3'). Keep in mind that although RNA incorporates Uracyl (U) instead of Thyamine (T), they are commonly used interchangeably when reading from sequence annotations. Some organisms have different codon tables (i.e. different codon-amino acid equivalency), but the same principle is followed.\n",
    "\n",
    "Across time, or due to mutagenic stress induced by environmental factors (e.g. ingested compounds, radiation, physical stress), mutations accumulate in the DNA of cells. Depending on the sites affected, mutations can either have no effect, alter the product of a gene, or alter the regulatory capacities of regions that allow for gene expression.\n",
    "\n",
    "![Alt Text](./Imgs/EffectMutations.png)\n",
    "> **Figure 3: Effect of mutations.** Mutations in regulatory elements can alter or completely inhibit gene expression, while mutations in genes can often render their product non-functional.\n",
    "\n",
    "Alterations within the codons of genes can benefit the fitness of an organism, and in turn become fixated over time within the population, or they could prevent the gene's product from functioning properly or completely. \n",
    "\n",
    "\n",
    "The degree to which a mutation affects a gene product is mainly determined by the particular site affected within the resulting **codons** of genes. If you pay close attention to *Figure 2*, you will notice that **a single amino acid can be encoded by multiple codons**. That is to say, there is a redundancy in the genetic code, generally referred to as *\"The Wobble Effect\"*, that has been selected for and conserved throughout billions of years of evolution. This redundancy then allows us to classify mutations as **synonymous** or **non-synonymous** based on the codon-amino acid relationship resulting from a Single Nucleotide Polymorphism (**SNP** - this is the common term denoting a mutation that has occurred at a single site).\n",
    "\n",
    "![Alt Text](./Imgs/SynonymousNonSynonymous.png)\n",
    "> **Figure 4: Synonymous VS Non-Synonymous mutation.** A single nucleotide polymorphism (SNP) in the codons of genes can result in an equivalent codon-amino acid pairing (i.e. the amino acid sequence remains the same, hence a *synonymous* mutation), or in a change in the  codon-amino acid pairing (i.e. a change in the amino acid sequence, hence a *non-synonymous* mutation).\n",
    "\n",
    "\n",
    "And that's it - we can start making useful tools with this backrgound given potential problems.\n",
    "\n",
    "## Annotation Systems.\n",
    "\n",
    "- **Fasta** : Contains the name/identifier of a gene or genomic region, followed by the nucleotide or amino acid sequence in said region. Identifiers and sequences can correspond to individual genes, chromosomes, whole genome scaffolds, etc. Files usually have '.fna' or '.fa' extensions.\n",
    "~~~~\n",
    ">NC_013993.1:c5651-5583 trnA (Denisova hominin tRNA)\n",
    "AAGGGCTTAGCTTAATTAAAGTGGCTGATTTGCGTTCAGTTGATGCAGAGTG\n",
    "GGGCTTTGCAGTCCTTA\n",
    "~~~~\n",
    "\n",
    "- **General Feature Format and General Tramsfer Format (GFF & GTF)** : Tab delimited file that contains information about features across genomes, including their name, location (chromosome: start-end), strand, and other comments.\n",
    "    - GTF Format: Chromosome | Genome Accession | Feature Type | Start | End | Score | Strand | something | Comments\n",
    "~~~~\n",
    "MT NC_013993.1  tRNA        5651 5583 . + . gene_id \"8923217\"; gene_name \"trnA\"; gene_source \"Denisova hominin\"; gene_biotype \"tRNA\"; \n",
    "~~~~\n",
    "\n",
    "- **Genbank** : File format created by the National Center of Biotechnology Information (NCBI). Comprises the most up to date annotations of individual genomes. Annotation format that possesses the information found in GFF/GTF files in addition to the sequences corresponding to said information. Bigger in size and thus harder to process. Click [here](https://www.ncbi.nlm.nih.gov/Sitemap/samplerecord.html) to see a Genbank annotation sample.\n",
    "\n",
    "\n",
    "- **BED** : Similar to the GFF/GTF file format, but has less required fields (only requires chromosome, start, and end of sequence).\n",
    "    - Format: Chromosome | Start | End || OPTIONAL: Name/ID | Score | Strand\n",
    "~~~~\n",
    "MT  5651   5583\n",
    "~~~~\n",
    "\n",
    "**PRACTICE 1.** Create simple tools that make use of the concepts learned in the 'Understanding Biology' section. We will be using a GTF annotation file containing all of E. coli's gene annotations, as well as fasta annotation file with *E. coli*'s complete genome. \n",
    "\n",
    "- Complement finder.\n",
    "> As mentioned before, DNA is a double-stranded molecule. We refer to one of the strands as the **positive strand** (+), and to the remaining strand as the **negative strand** (-). Genes may be found in either strand. We say that the positive and negative strands are **complementary** to each other given that both strands are held together by the hydrogen bond complementation of two nucleotide pairs, **C:G** & **A:T** respectively. \n",
    "![Alt Text](./Imgs/Complement.png)\n",
    "**Design a program that takes in a fasta file as input, but that outputs a fasta file with the complementary sequence instead. Note that when reading from fasta files, always assume that the sequence displayed corresponds to the positive (+) strand.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Write your code here\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Gene Extractor. \n",
    "> If we have a fasta file with *E. coli*'s complete genome, and a GTF file with all the genomic locations (chromosome:start-end) of genes, then we can use both files to generate a **multi-fasta** file containing all genes reported in *E. coli*'s genome. **Design a program that takes in a genomic fasta file and a gtf file as input, but that outputs a multi-fasta file containing all genes reported in *E. coli*'s genome.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Write your code here\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Auto-Translator. \n",
    "> Some organisms have different codon tables (i.e. the codon combinations for each amino acid differs). It would thus be ideal to write a program that can automatically translate nucleotide sequences in a multi-fasta file using a default codon table hard-coded into the program, that in turn also offers the user an option to input different codon tables from a file. **Create a program that takes in the multi-fasta file generated from the Gene Extractor tool - and optionally a file containing a different codon table - as input, but that outputs a multi-fasta file containing the protein-products of each amino acid sequence.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Write your code here.\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence alignments.\n",
    "\n",
    "\n",
    "### How do we obtain genome and gene sequences in the first place?\n",
    "Most [DNA/RNA sequencing](https://en.wikipedia.org/wiki/DNA_sequencing) methods, regardless of the techniques and equipment used, yield  **reads** in an annotation system called **fastq** that in turn contains an identifier, a quality value, and the sequence itself. The quality values (denoted as Q values) in fastq files act as a score that grades the probability of single nucleotides/bases to be correctly sequenced. For example, a Q value >= 20  in x base position in a read implies that there is less than 1/(10^2) chance of having an incorrect base call. This score helps the **read aligners** make decisions in regards to the validity of the alignments reported.\n",
    "\n",
    "Generally speaking, there are two types of **read aligners** : (i) **assemblers**, and (ii) **reference aligners**. For newly sequenced species or strains (i.e. no prior reference genome exists), we use **assemblers** to align reads resulting from **Whole Genome Sequencing (WGS)** projects between each other. Assemblers mix and match overlapping reads, slowly assembling the organism's genome like a puzzle. Genes are annotated based on **homology** found with genes derived from the genomes of other strains and species.\n",
    "\n",
    "![Alt Text](./Imgs/Shotgun_sequencing.jpg)\n",
    "> **Figure 5: Genome assembly scaffolding with the aid of an Assembler.** The most common sequencing technique used for WGS aimed at genome assembly is called *Shotgun Sequencing*, where the genome of an organism is broken down into uneaven fragments, and these fragments are then sequenced into reads in order to overlap them.\n",
    "\n",
    "Once an assembly and gene annotations become available for an organism, these are typically referred to as the *reference* genome/genes. We may then use **reference aligners** to align WGS or targeted-sequencing reads from other strains and species to the *reference* genome/genes. Below we will discuss different variants of reference aligners.\n",
    "\n",
    "### Search for homology (similarity between sequences).\n",
    "- **Basic Local Alignment Search Tool (BLAST)** : Fasta VS Database of sequences\n",
    "> Blast is the standard tool used to find homology, the existence of shared ancestry between a pair of structures, or genes, in different taxa. Asides from being able to find evolutionary relationships, homology studies are extremely useful to extrapolate gene and protein functions, metabolic routes, and other biological processes from annotated organisms to ones that have not yet been annotated. Blast can be accessed on [NCBI](https://blast.ncbi.nlm.nih.gov/Blast.cgi). There are several Blast iterations:\n",
    "    - **BLASTn** = Nucleotide VS Nucleotide\n",
    "    - **BLASTp** = Protein VS Protein (Most ideal for homology studies, given \"The Wobble Effect\")\n",
    "    - **BLASTt** = Protein VS Translated Nucleotides\n",
    "    - **BLASTx** = Translated nucleotides VS Proteins\n",
    "\n",
    "- **Magic Blast** : Fastq VS Database of Sequences\n",
    "> Magic blast is a Blast iteration that allows user to explore the content of raw reads by aligning them to a series of databases. This is particularly useful when you wish to verify the content of newly-sequenced reads. However, Magic blast has been mainly used in metagenomic studies, where the aim is to sequence and identify microorganisms found in environmental samples. Additionally, we often find a lot of reads within our projects that do not align to sequences found in the reference genome of our study subjects. To verify if this is due to contamination, we can use Magic blast to find potential matches in outside databases.\n",
    "\n",
    "**PRACTICE 2.** Go to [NCBI's Blast Page](https://blast.ncbi.nlm.nih.gov/Blast.cgi) and perform an alignment using the following protein sequence:\n",
    "~~~~\n",
    ">Unknown Seq\n",
    "MATKQLKPIFSTILRPFQNHRIRIIRFIVQFIMFGLLNGLIFGLARTQMLLPIAFPTGGPFSTVWNAFEA\n",
    "LQYLLTAWVFPYLALSVFMLFGSILGKTTCGWVCPFGLFQDLFRLIPIKKKRVSRPTNKSLSRIGTAILV\n",
    "FILIMSFIIGITFNNSGTKTTFGAGKDMPYSTIDPASTLFATLYYYLYWGMQNASFGAEIGQWKFLFFLR\n",
    "IIIFIIVIVLITLYPRAYCRWICPTGVLLGFFSRFSIFGLRVNKNRCTSGCDSCEKACPVQVPILTYDKD\n",
    "VTDKMCINCGECVDACKEGALKLTVRI\n",
    "~~~~\n",
    "\n",
    "All Blast algorithms serve as exploratory tools, but they do not provide the accuracy and efficiency that is necessary in order to analyze millions of mutations and isoform variants of genes. \n",
    "\n",
    "### Assessing mutations, calculating gene copy numbers, and determining gene expression.\n",
    "There is an extensive list of aligners capable of providing the resolution that is necessary for genome & transcriptome-wide studies, but [Bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml) still stands as the most widely used.\n",
    "> *\"Bowtie 2 is an ultrafast and memory-efficient tool for aligning sequencing reads to long reference sequences. It is particularly good at aligning reads of about 50 up to 100s or 1,000s of characters, and particularly good at aligning to relatively long (e.g. mammalian) genomes.\"*\n",
    "\n",
    "Aligners like Bowtie2 output an alignment file that contains the matches and mistmatches found in aligning reads within the context of the reference sequence(s) used. Below we show how a standard Bowtie2 pipeline looks like. Bioinformaticians use other tools, like the Samtools, GATK, and Picardtools suites, in order to curate results.\n",
    "\n",
    "~~~\n",
    "# A. Index genome to a Bowtie2 build (provides fast access to genome information)\n",
    "bowtie2-build GenomeAssembly.fna BuildName\n",
    "# B. Align sequenced data (reads) to build using Bowtie2. Outputs alignment file (.sam)\n",
    "bowtie2 -x BuildName -1 read_1 -2 read_2 --threads 16 -S BuildName.sam\n",
    "# C. Convert from sam format to bam format (bam is the binary version of sam -> occupies less space, fast random access)\n",
    "samtools view -bS BuildName.sam -@ 16 > BuildName.bam\n",
    "# D. Sort bam file concordantly using SAMtools\n",
    "samtools sort BuildName.bam BuildName.sorted -@ 16\n",
    "# E. Mark and remove duplicates in the alignment using Picardtools\n",
    "java -XX:ParallelGCThreads=16 -jar picard.jar MarkDuplicates INPUT=BuildName.sorted.bam OUTPUT=NoDup.bam METRICS_FILE=metrics.txt REMOVE_DUPLICATES=true VALIDATION_STRINGENCY=LENIENT\n",
    "# F. Define indel intervals using GATK (for realignment, seen in next step)\n",
    "java -XX:ParallelGCThreads=16 -jar GenomeAnalysisTK.jar -T RealignerTargetCreator -R Genome.fna -I NoDup.bam -o Intervals.txt\n",
    "# G. Realign against indels using GATK\n",
    "java -XX:ParallelGCThreads=16 -jar GenomeAnalysisTK.jar -T IndelRealigner -R Genome.fa -I NoDup.bam -targetIntervals Intervals.txt -o NoDupRealigned.bam\n",
    "# H. Call SNPs (variants) using SAMtools mpileup\n",
    "samtools mpileup -uf Genome.fa NoDupRealigned.bam | bcftools call -c > BuildName.vcf\n",
    "# I. Filter VCF based on BAQ (alignment Quality value - similar to the Q value of reads) or SNP depth using VCFtools\n",
    "vcftools --vcf BuildName.vcf --recode --recode-INFO-all --out BuildName --minQ 20.00 \n",
    "# J. Produce statistics for the VCF \n",
    "bcftools stats BuildName.recode.vcf > BuildNameFilteredVCFStats.txt\n",
    "~~~\n",
    "\n",
    "- **Assessing mutations.**\n",
    "> **Step H** uses information stored in the alignment files to identify **SNPs** in the sequenced reads that were used in the alignment process. This results in a Variant Call Format file, which can be analyzed downstream by user-made tools according to his/her needs. [SNPEff](http://snpeff.sourceforge.net/) is a good example of a user-made tool that became extremely popular. SNPEff evaluates SNPs annotated in VCFs based on their potential effect on the fitness of the organism. It takes into account the genomic features affected by the mutations (i.e. if the SNP is found in a protein coding gene, non-protein coding genes, regulatory regions, etc), and scores the potential impact of SNPs based on the biochemical properties resulting from amino acid changes. Mutations scattered across genomes can also be used to trace back population splits, and population history across time. By inputting a VCF file, coalescence models like [MSMC](https://www.nature.com/articles/ng.3015) can output population histories that are of great use to archaeologists, historians, anthropologists, and biologists alike.\n",
    "\n",
    "\n",
    "- **Calculating gene copy numbers.**\n",
    "> We can use already annotated gene sequences in order to estimate how many copies of a single gene are present within the genome of an organism. This can be done through a read coverage-based approach. **Read coverage** is calculated in this case as the percentage of the gene region covered by reads out of the total gene length. The more reads that align to a gene reference, the higher the coverage. To calculate coverage from an alignment file, one can simply use Samtools, followed by an awk command for the arithmetic: \n",
    "~~~\n",
    "samtools depth BuildName.sorted.bam | awk '{sum+=$3; sumsq+=$3*$3} END { print \"Average = \",sum/NR; print \"Stdev = \",sqrt(sumsq/NR - (sum/NR)**2)}' > Coverage.txt\n",
    "~~~\n",
    "The figure below illustrates the workflow utilized to estimate gene copies.\n",
    "![Alt Text](./Imgs/MammalCopyNumber.png)\n",
    ">**Figure 6: Workflow used to calculate gene copy numbers in mammalian species.** Initial Blastp alignments between well-annotated genes (query - in this case human genes) and the proteome of the organism under study (Database) are carried out with the purpose of identifying and quantifying already-annotated gene copies. Unnanotated copies are identified through a coverage based approach: raw reads are aligned to a multi-fasta file containing gene(s) under study, and the coverage of each individual gene is divided by the average coverage of all genes (most genes tend to have a single copy, so the average coverage of all genes is representative of the coverage value for a single copy). Resulting copy numbers are then compared, and a decision boundary is implemented.\n",
    "\n",
    "\n",
    "- **Determining gene expression.**\n",
    ">While all cells of multicellular organisms share the same genome, each and every one of them have different gene expression profiles - this is called **differential gene expression**. However, if we review *The Central Dogma*, we quickly realize that there are two levels of gene expression: (i) **transcription**, and (ii) **translation**. To evaluate gene expression at the level of **transcription**, we use a type of sequencing technology generally known as **RNA-Seq**. There are different types of RNA-Seq techniques, but the overarching goal of these techniques is to isolate and sequence sub-pools of gene-derived RNAs. To evaluate gene expression at the level of **translation**, we use a sequencing approach known as **Ribosome profiling**. \n",
    "![Alt Text](./Imgs/RibosomeProfiling.png)\n",
    "> **Figure 7: Schematic of ribosome profiling procedure.** Ribosome profiling adds an extra step to standard RNA-Seq protocols: samples are treated with a reactant that degrades RNA in the cytosol of cells, leaving RNAs found within ribosomes - the cell's translational apparatus - free of harm. These ribosome-protected RNAs are then collected and sequenced. Source of Figure: [Wikipedia](https://en.wikipedia.org/wiki/Ribosome_profiling). Note that in both RNA-Seq and Ribosome profiling, we can design short complementary sequences known as primers to select for, and amplify certain sub-pools of RNA when sequencing.\n",
    "\n",
    "Instead of simply using Bowtie2, most differential expression projects are carried out using a special tool that makes use of Bowtie2, called [TopHat](https://ccb.jhu.edu/software/tophat/index.shtml), which takes into account splicing variations of genes when dealing with eukaryotic genomes. [Cufflinks](http://cole-trapnell-lab.github.io/cufflinks/) is commonly used in conjunction with TopHat in order to calculate overall gene expression. We include an example of the standard pipeline below:\n",
    "~~~\n",
    "#Mandatory step for ribosome profiling reads:\n",
    "Trim length of reads to 23 nucleotides each. Not necessary for RNA-Seq reads.\n",
    "\n",
    "#Alignment step (once per sample)\n",
    "tophat2 --GTF genes.gtf --transcriptome-index Transcriptome/genes -o TopHatOut -p 16 --no-novel-juncs -g 1 GenomeAssembly.fa readNames\n",
    "\n",
    "#Differential expression of two or more samples\n",
    "cuffdiff -p 16 -b GenomeAssembly.fa -o cuffdiffOutput genes.gtf TopHatOut/accepted_hits.bam TopHatOut_2/accepted_hits.bam\n",
    "~~~\n",
    "\n",
    "We did not discuss other mainstream qualitative and quantitative bioinformatic approaches, like [micro-arrays](https://en.wikipedia.org/wiki/DNA_microarray) and [mass spectrometry](https://en.wikipedia.org/wiki/Mass_spectrometry), but we encourage students to read about them and think about their strengths and weaknesses compared to sequencing technology.\n",
    "\n",
    "\n",
    "## Visualizing sequences.\n",
    "- **Genomes content and gene expression.**\n",
    "\n",
    "A Bioinformatician's job is not only to analyze biological data, but also to visualize and represent it in a way that readers can understand. This section will serve as a guide to illustrate common visualization tasks in the field of Bioinformatics.\n",
    "\n",
    "\n",
    "![Alt Text](./Imgs/GenomePlot.svg)\n",
    "> **Figure 8: Content of bacterial genome.** The genomes of bacteria are circular in shape, and so it is convenient to represent genic and intergenic regions in the genome in said shape. The outer circle represents the positive strand, and the inner circle represents the negative strand. Each genomic region is assigned an angle based on the notion that angle = (nucleotide_count/genome_length)*360. The  code for this plot is available [here](https://gist.github.com/CharlesSanfiorenzo/a8e86755114cec7264996a12548272dc).\n",
    "\n",
    "\n",
    "![Alt Text](./Imgs/RiboBind.png)\n",
    "> **Figure 9: Alignment score reported for a regulatory motif found in RNA VS overall gene expression.** In order to uncover the regulatory potential of motifs found in DNA or RNA, we can create alignments between sequences under study and well-characterized regulatory motifs. The resulting alignment scores can then be correlated to expression levels reported - potentially serving as training sets for linear regression models or Support Vector Machines.\n",
    "\n",
    "\n",
    "![Alt Text](./Imgs/VolcanoPlot.png)\n",
    "> **Figure 10: Volcano plot of differentially expressed genes.** There are hundreds-thousands of genes in the genomes of multicellular organisms, most of which are differentially expressed in different cell types to a certain degree. We can score genes using p values, or other statistical tests, to later establish an alpha value in order to determine what genes are differentially expressed in accordance to statistical significance. This selection process can be represented by a volcano plot, as seen above.\n",
    "\n",
    "![Alt Text](./Imgs/HeatMap.png)\n",
    "> **Figure 11: Heat map of selected differentially expressed genes in CD34+ and CD34- cells.** After selecting genes based on statistical significance and magnitude of differential expression, we can group and show the expression patterns of said genes with efficacy. We can also group genes and present statistics based on their biological functions - typically done using Gene Ontology (GO) accessions of genes (not shown).\n",
    "\n",
    "- **RNA-RNA interactions.**\n",
    "\n",
    "Although RNA is single stranded, the nucleotides in RNA possess the same complementarity as the nucleotides in DNA (i.e. **C:G** & **A:U/T** respectively). Given favorable thermodynamic conditions, RNAs tend to form what is referred to as secondary structures throughout complementation of nucleotides along a single strand; this is a *cis* interaction. RNAs can also interact with other RNAs (a *trans* interaction). An **RNA Interactome** describes all the possible  cis/trans RNA-RNA interactions documented by recent sequencing approaches (namely [MARIO](https://www.nature.com/articles/ncomms12023) datasets). Given their difficulty to process and interpret (we're assuming that you don't know your linear algebra yet), it is easier to explore these sequencing approaches through online tools. We recommend using [RISE](http://rise.life.tsinghua.edu.cn/).\n",
    "\n",
    "![Alt Text](./Imgs/RNAInteractome_Ed.png)\n",
    "> **Figure 12: RNA Interactome for the ribosome encoding gene, RPL21.**\n",
    "\n",
    "\n",
    "\n",
    "Given the biological importance of these secondary structures, many mathematical models have been designed with the sole purpose of predicting secondary structures of RNAs, provided we know their nucleotide sequence. [ViennaRNA](https://www.tbi.univie.ac.at/RNA/) is a great tool suite designed to predict the secondary structures of different types of RNAs. The figures below were produced using RNASnoop - a tool included in the ViennaRNA suite. If you're interested in designing your own structure prediction algorithms, the authors behind ViennaRNA have put together an excellent [how-to manual](http://www.bioinf.uni-leipzig.de/Leere/SS15/Bioinf2/lengauer.pdf) on secondary structures of RNA.\n",
    "\n",
    "\n",
    "![Alt Text](./Imgs/RNARNAInteraction_Ed.png)\n",
    "> **Figure 13: cis/trans interactions of a snoRNA with several target sequences.**\n",
    "\n",
    "\n",
    "- **Phylogeny.**\n",
    "\n",
    "Phylogeny is the study of the evolutionary history and relationships among groups of organisms (e.g. species & populations). Instead of depending on morphological traits, we can use genomics to calculate evolutionary distance and create phylogenetic trees.\n",
    "\n",
    "\n",
    "![Alt Text](./Imgs/AsgardPhylogeny.jpg)\n",
    "> **Figure 14: Asgard archae - newly discovered domain of life through the employment of metagenomics and phylogenetics.** Asgard archae are currently considered to be the closest relatives to modern eukaryotes. Read [more](https://www.nature.com/articles/nature21031).\n",
    "\n",
    "Phylogenetic and Phylogenomic trees are created through a series of algorithms that rely on different pair-wise distance alignments, learning models, and scoring matrices to yield the tree with the most likely topology. We recommend using [MEGA](https://www.megasoftware.net/) to play around and create your first phylogenetic trees. However, MEGA is not recommended for publishable work - we recommend using [MrBayes](http://mrbayes.sourceforge.net/) instead.\n",
    "\n",
    "\n",
    "## Useful links.\n",
    "- [NCBI](https://www.ncbi.nlm.nih.gov/) - Assemblies, genes, annotation files. Manual search, or visit their ftp page.\n",
    "- [NCBI's SRA]() - Publicly available reads\n",
    "- [NCBI dbSNP](https://www.ncbi.nlm.nih.gov/snp/) - Database of SNPs reported across thousands of genomes (EXTREMELY USEFUL FOR CLINICAL STUDIES).\n",
    "- [NCBI's Gene Expression Omnibus (GEO)](https://www.ncbi.nlm.nih.gov/geo/) - Reads pertaining gene expression projects\n",
    "- [European Nucleotide Archive (ENA)](https://www.ebi.ac.uk/ena) - Publicly available reads, no need to convert SRA -> Fastq\n",
    "\n",
    "\n",
    "***\n",
    "#### Course authors: \n",
    "- [Charles Sanfiorenzo](https://github.com/CharlesSanfiorenzo/Bioinformatics) - csanfior@caltech.edu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
